rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,98 @@
+# RFC 5843 - 2021-01-08 - Encoding/Decoding for Vector to Vector Communication
+
+There has been an ongoing discussion of what format Vector-to-Vector communications should take. For some examples of issues related to the subject check out #5809, #5124, or #5341. This RFC is intended to propose a format for standardization of our vector-to-vector communications.
+
+## Scope
+
+- This RFC should cover the encoding format we intend to leverage
+- It should also cover any alterations that might need to be made to the expected data format
+
+## Motivation
+
+The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.
+
+## Internal Proposal
+
+From a high level my proposal is pretty simple. I'll dive into the rationale and context lower in the doc but lets just tee that up, shall we?
+
+I propose that if we're not explicitly searching for alternate tooling because we've decided that we wholly can't deal with the tooling overhead of protobufs, we should instead continue to use them with prost. As an added dimension to the issue I'd suggest that we support an initial Transport of HTTP/1.1 and consider a fast-follow of effort to implement HTTP/2 (or even optionally GRPC - though it bares saying I am unconvinced of the necessity of GRPC over HTTP/2 for our usecase).
```

Comment:
Could we go into a little more detail about how we want to marry protobuf with HTTP? Right now we just have protobuf definitions for individual events, which we certainly aren't going to send one at a time each in their own request, so we will need some way to batch up those individual messages. Do you think we should define a new protobuf type for batches of events or similar?

I think the argument in favor of gRPC is less about the additional features it would provide and more about not rolling our own implementation of something that we could just take off the shelf. I'm also not convinced it's worth it, but I do think it's worth investigating to some level of depth so that we can make an informed decision. If it's as simple as using `tonic` and implementing a trait or two, that could be a compelling option over doing our own manual protobuf over HTTP (especially given features like streaming that we may want to implement in the future). Ideally these are the kinds of unknowns and future risks that we can try to suss out at the RFC phase, before we're already down the road of implementation :smile:

Reply:
I agree, it is important to understand the trade-offs. I think we should investigate how do the gRPC and of the HTTP/2 compare with, say, commonly known TCP and HTTP/1. Also, HTTP/3 is a thing - we might want to consider it too.